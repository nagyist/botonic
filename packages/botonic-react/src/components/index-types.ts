import React from 'react'

import { SENDERS } from '../index-types'

export type MessageType =
  | 'audio'
  | 'buttonmessage'
  | 'carousel'
  | 'custom'
  | 'document'
  | 'image'
  | 'location'
  | 'text'
  | 'video'

export interface MessageProps {
  blob?: boolean
  children?: React.ReactNode
  delay?: number
  enabletimestamps?: boolean
  sentBy?: SENDERS
  json?: Record<string, unknown>
  style?: Record<string, unknown>
  type?: MessageType
  typing?: number
}

export interface TextProps extends MessageProps {
  // converts markdown syntax to HTML
  markdown?: boolean
  feedbackEnabled?: boolean
  inferenceId?: string
  botInteractionId?: string
}

export interface Webview {
  name: string
}

export interface ButtonProps {
  params?: any
  path?: string
  payload?: string
  target?: string
  url?: string
  webview?: Webview
  onClick?: () => void
  autodisable?: boolean
  disabled?: boolean
  disabledstyle?: boolean
  children: string
  setDisabled?: (disabled: boolean) => void
  parentId?: string
}

export interface ReplyProps {
  path?: string
  payload?: string
  children: string
}

export interface PicProps {
  src: string
}

export interface ImageProps extends MessageProps {
  src: string
  input?: { data: string }
}

export interface VideoProps extends MessageProps {
  src: string
  input?: { data: string }
}

export interface AudioProps extends MessageProps {
  src: string
  input?: { data: string }
}

export interface DocumentProps extends MessageProps {
  src: string
  input?: { data: string }
  from?: any
}

export interface TitleProps {
  children: React.ReactNode
  style: string
}

export type SubtitleProps = TitleProps

export interface CustomMessageType {
  (props: any): JSX.Element
  customTypeName: string
  deserialize(msg: any): JSX.Element
}

export type BlockInputOption = {
  preprocess?: (message: string) => string
  match: RegExp[]
  message: string
}

export type WrappedComponent<Props> = React.FunctionComponent<Props> & {
  customTypeName: string
}
