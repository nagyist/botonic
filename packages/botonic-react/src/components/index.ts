export { Audio } from './audio'
export { Button } from './button'
export { Carousel } from './carousel'
export { customMessage } from './custom-message'
export { Document } from './document'
export { Element } from './element'
export { Handoff } from './handoff'
export { Image } from './image'
export * from './index-types'
export { Location } from './location'
export { Message } from './message'
export { MessageTemplate } from './message-template'
export * from './multichannel'
export { Pic } from './pic'
export { Raw } from './raw'
export { Reply } from './reply'
export { ShareButton } from './share-button'
export { Subtitle } from './subtitle'
export { Text } from './text'
export { Title } from './title'
export { Video } from './video'
export {
  normalizeWebchatSettings,
  WebchatSettings,
  WebchatSettingsProps,
} from './webchat-settings'
export {
  WhatsappButtonList,
  WhatsappButtonListProps,
  WhatsappButtonListRowProps,
  WhatsappButtonListSectionProps,
} from './whatsapp-button-list'
export { WhatsappCatalog, WhatsappCatalogProps } from './whatsapp-catalog'
export {
  WhatsappCTAUrlButton,
  WhatsappCTAUrlButtonProps,
} from './whatsapp-cta-url-button'
export {
  WhatsappMediaCarousel,
  WhatsappMediaCarouselProps,
} from './whatsapp-media-carousel'
export { WhatsappProduct } from './whatsapp-product'
export {
  WhatsappProductCarousel,
  WhatsappProductCarouselProps,
} from './whatsapp-product-carousel'
export {
  ProductItem,
  WhatsappProductList,
  WhatsappProductListProps,
  WhatsappProductListSection,
} from './whatsapp-product-list'
export { WhatsappTemplate } from './whatsapp-template'
