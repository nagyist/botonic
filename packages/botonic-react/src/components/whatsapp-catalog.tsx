import { INPUT } from '@botonic/core'
import React from 'react'

import { renderComponent } from '../util/react'
import { Message } from './message'

export interface WhatsappCatalogProps {
  body: string
  footer?: string
  thumbnailProductId?: string
}

const serialize = (message: string) => {
  return { text: message }
}

export const WhatsappCatalog = (props: WhatsappCatalogProps) => {
  const renderBrowser = () => {
    // Return a dummy message for browser
    const message = `WhatsApp Catalog would be sent to the user.`
    return (
      <Message json={serialize(message)} {...props} type={INPUT.TEXT}>
        {message}
      </Message>
    )
  }

  const renderNode = () => {
    return (
      // @ts-ignore Property 'message' does not exist on type 'JSX.IntrinsicElements'.
      <message
        {...props}
        body={props.body}
        footer={props.footer}
        thumbnailProductId={props.thumbnailProductId}
        type={INPUT.WHATSAPP_CATALOG}
      />
    )
  }

  return renderComponent({ renderBrowser, renderNode })
}
