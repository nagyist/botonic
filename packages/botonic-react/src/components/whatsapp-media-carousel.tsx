import { INPUT } from '@botonic/core'
import React from 'react'

import { toSnakeCaseKeys } from '../util/functional'
import { renderComponent } from '../util/react'
import { Message } from './message'

type Parameters = TextParameter | CurrencyParameter | DateTimeParameter

interface TextParameter {
  type: 'text'
  text: string
}

interface CurrencyParameter {
  type: 'currency'
  currency: {
    fallbackValue: string
    code: string
    amount1000: number
  }
}

interface DateTimeParameter {
  type: 'date_time'
  dateTime: { fallbackValue: string }
}

type CardButton = QuickReplyButton | UrlButton

interface Button {
  type: 'quick_reply' | 'url'
  buttonIndex?: number
}

interface QuickReplyButton extends Button {
  payload: string
}

interface UrlButton extends Button {
  urlVariable: string
}

interface Card {
  fileType: 'image' | 'video'
  fileId: string
  cardIndex?: number
  bodyParameters?: Parameters[]
  buttons?: CardButton[]
  extraComponents?: Record<string, any>[]
}

export interface WhatsappMediaCarouselProps {
  templateName: string
  templateLanguage: string
  cards: Card[]
  bodyParameters?: Parameters[]
}

const serialize = (message: string) => {
  return { text: message }
}

export const WhatsappMediaCarousel = (props: WhatsappMediaCarouselProps) => {
  const renderBrowser = () => {
    // Return a dummy message for browser
    const message = `WhatsApp Media Carousel would be sent to the user.`
    return (
      <Message json={serialize(message)} {...props} type={INPUT.TEXT}>
        {message}
      </Message>
    )
  }

  const getCards = (cards: Card[]) => {
    cards.forEach((card, index) => {
      if (!card.cardIndex) {
        card.cardIndex = index
      }
      card.buttons?.forEach((button, index) => {
        if (!button.buttonIndex) {
          button.buttonIndex = index
        }
      })
    })
    return toSnakeCaseKeys(cards)
  }

  const renderNode = () => {
    return (
      // @ts-ignore Property 'message' does not exist on type 'JSX.IntrinsicElements'.
      <message
        {...props}
        bodyParameters={JSON.stringify(toSnakeCaseKeys(props.bodyParameters))}
        cards={JSON.stringify(getCards(props.cards))}
        templateName={props.templateName}
        templateLanguage={props.templateLanguage}
        type={INPUT.WHATSAPP_MEDIA_CAROUSEL}
      />
    )
  }

  return renderComponent({ renderBrowser, renderNode })
}
