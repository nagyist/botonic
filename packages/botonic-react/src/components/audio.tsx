import { INPUT, isBrowser } from '@botonic/core'
import React from 'react'
import styled from 'styled-components'

import { ROLES } from '../constants'
import { staticAsset } from '../util/environment'
import { AudioProps } from './index-types'
import { Message } from './message'

const StyledAudio = styled.audio`
  max-width: 100%;
`

const serialize = (audioProps: { src: string }) => {
  return { audio: audioProps.src }
}

export const Audio = (props: AudioProps) => {
  props = { ...props, src: staticAsset(props.src) }
  let content = props.children
  if (isBrowser())
    content = (
      <StyledAudio id='myAudio' controls>
        <source src={props.src} type='audio/mpeg' />
        Your browser does not support this audio format.
      </StyledAudio>
    )
  return (
    <Message
      role={ROLES.AUDIO_MESSAGE}
      json={serialize(props)}
      {...props}
      type={INPUT.AUDIO}
    >
      {content}
    </Message>
  )
}

Audio.serialize = serialize
