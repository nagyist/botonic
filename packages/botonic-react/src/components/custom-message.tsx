import { INPUT } from '@botonic/core'
import merge from 'lodash.merge'
import React from 'react'

import { SENDERS } from '../index-types'
import { createErrorBoundary } from '../util/error-boundary'
import { warnDeprecatedProps } from '../util/logs'
import { mapObjectNonBooleanValues } from '../util/react'
import { CustomMessageType } from './index-types'
import { Message } from './message'
import { Reply } from './reply'

/**
 *
 *  name: as it appears at ThemeProps' message.customTypes key
 *  CustomMessageComponent
 *  defaultProps: Props for the wrapper Message
 *  ErrorBoundary: to recover in case it fails
 */

export interface CustomMessageArgs {
  name: string
  component: React.ComponentType
  defaultProps?: Record<string, any>
  errorBoundary?: any
}

export const customMessage = ({
  name,
  component: CustomMessageComponent,
  defaultProps = {},
  errorBoundary = createErrorBoundary(),
}: CustomMessageArgs): CustomMessageType => {
  const CustomMessage = props => {
    warnDeprecatedProps(defaultProps, 'customMessage:')
    if (defaultProps.sentBy === SENDERS.user) {
      defaultProps.ack = 1
    }
    return (
      <Message
        {...merge(mapObjectNonBooleanValues(defaultProps), props)}
        type={INPUT.CUSTOM}
      />
    )
  }

  const splitChildren = props => {
    const { children } = props
    const isReply = e => e.type === Reply
    try {
      if (!Array.isArray(children) && !isReply(children)) {
        return { replies: null, childrenWithoutReplies: children }
      }
      const childrenArray = React.Children.toArray(children)
      const replies = childrenArray.filter(isReply)
      const childrenWithoutReplies = childrenArray.filter(e => !isReply(e))
      return {
        replies: replies,
        childrenWithoutReplies,
      }
    } catch (e) {
      return { replies: null, childrenWithoutReplies: children }
    }
  }

  const WrappedComponent = props => {
    const { id, children, ...customMessageProps } = props
    const { replies, childrenWithoutReplies } = splitChildren(props)

    const ErrorBoundary = errorBoundary

    return (
      <CustomMessage
        id={id}
        json={{
          ...customMessageProps,
          id,
          children: childrenWithoutReplies,
          customTypeName: name,
        }}
        sentBy={props.sentBy || SENDERS.bot}
        isUnread={props.isUnread}
      >
        <ErrorBoundary key={'errorBoundary'} {...customMessageProps}>
          <CustomMessageComponent {...customMessageProps}>
            {childrenWithoutReplies}
          </CustomMessageComponent>
        </ErrorBoundary>
        {replies}
      </CustomMessage>
    )
  }
  WrappedComponent.customTypeName = name
  // eslint-disable-next-line react/display-name
  WrappedComponent.deserialize = (msg: any) => (
    <WrappedComponent
      id={msg.id}
      key={msg.key}
      json={msg.data}
      {...msg.data}
      sentBy={msg.sentBy || SENDERS.bot}
      isUnread={msg.isUnread}
    />
  )
  return WrappedComponent
}
