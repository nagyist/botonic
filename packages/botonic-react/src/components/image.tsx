import { INPUT, isBrowser } from '@botonic/core'
import React, { useContext, useState } from 'react'
import styled from 'styled-components'

import { ROLES, WEBCHAT } from '../constants'
import { WebchatContext } from '../contexts'
import { staticAsset } from '../util/environment'
import { ImageProps } from './index-types'
import { Message } from './message'

interface StyledImageProps {
  hasPreviewer: boolean
}

const StyledImage = styled.img<StyledImageProps>`
  border-radius: 8px;
  max-width: 150px;
  max-height: 150px;
  margin: -3px -6px;
  cursor: ${({ hasPreviewer }) => (hasPreviewer ? 'pointer' : 'auto')};
`

const serialize = (imageProps: { src: string }) => {
  return { image: imageProps.src }
}

export const Image = (props: ImageProps) => {
  props = props.input?.data
    ? { ...props, src: props.input.data }
    : { ...props, src: staticAsset(props.src) }

  let content = props.children

  const [isPreviewerOpened, setIsPreviewerOpened] = useState(false)
  const openPreviewer = () => setIsPreviewerOpened(true)
  const closePreviewer = () => setIsPreviewerOpened(false)

  const { getThemeProperty } = useContext(WebchatContext)
  const ImagePreviewer = getThemeProperty(
    WEBCHAT.CUSTOM_PROPERTIES.imagePreviewer,
    null
  )
  if (isBrowser()) {
    content = (
      <>
        <StyledImage
          src={props.src}
          onClick={openPreviewer}
          hasPreviewer={Boolean(ImagePreviewer)}
        />
        {ImagePreviewer && (
          <ImagePreviewer
            src={props.src}
            isPreviewerOpened={isPreviewerOpened}
            openPreviewer={openPreviewer}
            closePreviewer={closePreviewer}
          />
        )}
      </>
    )
  }
  return (
    <Message
      role={ROLES.IMAGE_MESSAGE}
      json={serialize(props)}
      {...props}
      type={INPUT.IMAGE}
    >
      {content}
    </Message>
  )
}

Image.serialize = serialize
