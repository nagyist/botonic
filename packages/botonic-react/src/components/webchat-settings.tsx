import { INPUT } from '@botonic/core'
import React, { useContext } from 'react'

import { renderComponent } from '../util/react'
import { stringifyWithRegexs } from '../util/regexs'
import { WebchatContext } from '../webchat/context'
import { PersistentMenuOptionsTheme, ThemeProps } from '../webchat/theme/types'
import { BlockInputOption } from './index-types'

export interface WebchatSettingsProps {
  blockInputs?: BlockInputOption[]
  enableAnimations?: boolean
  enableAttachments?: boolean
  enableEmojiPicker?: boolean
  enableUserInput?: boolean
  persistentMenu?: PersistentMenuOptionsTheme
  theme?: ThemeProps
  user?: { extra_data?: any }
}

export const WebchatSettings = ({
  theme,
  blockInputs,
  persistentMenu,
  enableEmojiPicker,
  enableAttachments,
  enableUserInput,
  enableAnimations,
  user,
}: WebchatSettingsProps) => {
  const renderBrowser = () => {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const { updateWebchatDevSettings } = useContext(WebchatContext)
    updateWebchatDevSettings({
      theme,
      blockInputs,
      persistentMenu,
      enableEmojiPicker,
      enableAttachments,
      enableUserInput,
      enableAnimations,
    })
    return null
  }
  const renderNode = () => {
    const updatedTheme = normalizeWebchatSettings({
      theme,
      blockInputs,
      persistentMenu,
      enableEmojiPicker,
      enableAttachments,
      enableUserInput,
      enableAnimations,
    })
    return (
      //@ts-ignore
      <message
        type={INPUT.WEBCHAT_SETTINGS}
        settings={stringifyWithRegexs({ theme: updatedTheme, user })}
      />
    )
  }
  return renderComponent({ renderBrowser, renderNode })
}

export const normalizeWebchatSettings = (settings: WebchatSettingsProps) => {
  let {
    theme,
    blockInputs,
    persistentMenu,
    enableEmojiPicker,
    enableAttachments,
    enableUserInput,
    enableAnimations,
  } = settings
  if (!theme) theme = {}
  if (!theme.userInput) theme.userInput = {}
  if (persistentMenu !== undefined) {
    theme.userInput.persistentMenu = persistentMenu
  }
  if (enableEmojiPicker !== undefined) {
    if (!theme.userInput.emojiPicker) theme.userInput.emojiPicker = {}
    theme.userInput.emojiPicker.enable = enableEmojiPicker
  }
  if (enableAttachments !== undefined) {
    if (!theme.userInput.attachments) theme.userInput.attachments = {}
    theme.userInput.attachments.enable = enableAttachments
  }
  if (enableUserInput !== undefined) {
    theme.userInput.enable = enableUserInput
  }
  if (blockInputs !== undefined) theme.userInput.blockInputs = blockInputs

  if (!theme.animations) theme.animations = {}
  if (enableAnimations !== undefined) {
    theme.animations.enable = enableAnimations
  }
  return theme
}
