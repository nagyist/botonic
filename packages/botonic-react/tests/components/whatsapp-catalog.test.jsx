import { expect, test } from '@jest/globals'
import React from 'react'
import TestRenderer from 'react-test-renderer'

import { WhatsappCatalog } from '../../src/components'

const renderToJSON = sut => TestRenderer.create(sut).toJSON()

test('renders Whatsapp Catalog component', () => {
  const props = {
    body: 'Catalog sent from test',
    thumbnailProductId: 'fake-id',
    footer: 'Catalog footer',
  }

  const tree = renderToJSON(<WhatsappCatalog {...props} />)
  expect(tree).toMatchSnapshot()
})
