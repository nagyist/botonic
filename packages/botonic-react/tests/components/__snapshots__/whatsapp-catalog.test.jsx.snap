// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Whatsapp Catalog component 1`] = `
<message
  body="Catalog sent from test"
  footer="Catalog footer"
  thumbnailProductId="fake-id"
  type="whatsapp-catalog"
/>
`;
